lib/Service/ScrobblerService.php

```diff
@@ -0,0 +1,234 @@
+		$this->urlGenerator = $urlGenerator;
+		$this->trackBusinessLayer = $trackBusinessLayer;
+		$this->crypto = $crypto;
+		$this->appName = $appName;
+	}
+
+	/**
+	 * @throws \Throwable when unable to generate or save a session
+	 */
+	public function generateSession(string $token, string $userId) : void {
+		$scrobbleService = $this->getApiService();
+		$ch = $this->makeCurlHandle($scrobbleService);
+		$params = $this->generateBaseMethodParams('auth.getSession');
+		$params['token'] = $token;
+		$params['api_sig'] = $this->generateSignature($params);
+		\curl_setopt($ch, \CURLOPT_POSTFIELDS, \http_build_query($params));
+		$sessionText = \curl_exec($ch);
+		$xml = \simplexml_load_string($sessionText);
+
+		$status = (string)$xml['status'];
+		if ($status !== 'ok') {
+			throw new \Exception((string)$xml->error, (int)$xml->error['code']);
+		}
+
+		try {
+			$encryptedKey = $this->crypto->encrypt(
+				(string)$xml->session->key,
+				$userId . $this->config->getSystemValue('secret')
+			);
+			$this->config->setUserValue($userId, $this->appName, 'scrobbleSessionKey', $encryptedKey);
+		} catch (\Throwable $e) {
+			$this->logger->error('Unable to save session key ' . $e->getMessage());
+			throw $e;
+		}
+	}
+
+	public function getApiKey() : ?string {
+		return $this->config->getSystemValue('music.scrobble_api_key', null);
+	}
+
+	public function getApiSecret() : ?string {
+		return $this->config->getSystemValue('music.scrobble_api_secret', null);
+	}
+
+	public function getApiService() : ?string {
+		return $this->config->getSystemValue('music.scrobble_api_service', null);
+	}
+
+	public function getApiSession(string $userId): ?string
+	{
+		$encryptedKey = $this->config->getUserValue($userId, $this->appName, 'scrobbleSessionKey');
+		if (!$encryptedKey) {
+			return null;
+		}
+		$key = $this->crypto->decrypt($encryptedKey, $userId . $this->config->getSystemValue('secret'));
+		return $key;
+	}
+
+	public function getTokenRequestUrl(): ?string {
+		$apiKey = $this->getApiKey();
+		$apiService = $this->getApiService();
+		if (!$apiKey || !$apiService) {
+			return null;
+		}
+
+		$tokenHandleUrl = $this->urlGenerator->linkToRouteAbsolute('music.scrobbler.handleToken');
+		switch ($apiService) {
+			case 'lastfm':
+				return "http://www.last.fm/api/auth/?api_key={$apiKey}&cb={$tokenHandleUrl}";
+			default:
+				throw new \Exception('Invalid service');
+		}
+	}
+
+	/**
+	 * @param array<int,mixed> $trackIds
+	 */
+	public function scrobbleTrack(array $trackIds, string $userId, \DateTime $timeOfPlay) : bool {
+		$sessionKey = $this->getApiSession($userId);
+		$scrobbleService = $this->getApiService();
+		if (!$sessionKey || !$scrobbleService) {
+			return false;
+		}
+
+		$timestamp = $timeOfPlay->getTimestamp();
+		$scrobbleData = \array_merge($this->generateBaseMethodParams('track.scrobble'), [
+			'sk' => $sessionKey,
+		]);
+
+		/** @var array<Track> $tracks */
+		$tracks = $this->trackBusinessLayer->findById($trackIds);
```

Comment:
Also `userId` should be passed to `findById`. Otherwise we may leak one user's data to other users of the same cloud.